### PR TITLE
Add Email Tools and MSSQL PSA service

### PR DESCRIPTION
**Pull request by aminelouati**

### CODE REVIEW

File: DataConnectors/resources/catalog/SQL_Pooled_Connection_Query.xml [hunk 1]
```diff
@@ -6,11 +6,13 @@
     <variable name="RDBMS_NAME" value="hsqldb" model="PA:LIST(postgresql, mysql, greenplum, sqlserver, oracle, hsqldb)" group="Database Connection" description="The relational database management system name e.g. postgresql, mysql, greenplum, hsqldb, oracle..."/>
     <variable name="DATABASE" value="scheduler" model="PA:NOT_EMPTY_STRING" group="Database Connection" description="The database name"/>
     <variable name="HOST" value="localhost" model="PA:NOT_EMPTY_STRING" group="Database Connection" description="Hostname or IP address of the database host machine. Default value is localhost"/>
-    <variable name="PORT" value="" model="PA:INTEGER?" group="Database Connection" description="The listening port of the database"/>
+    <variable name="PORT" value="9001" model="PA:INTEGER" group="Database Connection" description="The listening port of the database"/>
```
Comment (fviale) [inline DataConnectors/resources/catalog/SQL_Pooled_Connection_Query.xml:9]: I think this parameter was optional because when not provided, the workflow will use the default port of the chosen database, why did you change this ?

Comment (aminelouati) [reply DataConnectors/resources/catalog/SQL_Pooled_Connection_Query.xml:9]: If left empty, the credential key will not include a port when the user enters the password in the format ${RDBMS_NAME}://${USERNAME}@${HOST}:${PORT}. However, the code will assign a default value to this empty field. As a result, credentials.get will return a null value.

Comment (aminelouati) [reply DataConnectors/resources/catalog/SQL_Pooled_Connection_Query.xml:9]: To prevent this issue, it is recommended to require the user to enter the port.

Comment (fviale) [reply DataConnectors/resources/catalog/SQL_Pooled_Connection_Query.xml:9]: I see, does it work with an inferred variable ?

File: EmailTools/resources/catalog/Imap_Email_Checker.xml [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    <variable advanced="false" description="Email user name" group="SMTP Server Connection" hidden="false" model="PA:NOT_EMPTY_STRING" name="USERNAME" value="louati.amiine@gmail.com"/>
+    <variable advanced="false" description="Email app password" group="SMTP Server Connection" hidden="false" model="PA:CREDENTIAL" name="PASSWORD" value="imap://${USERNAME}@${HOST}"/>
+    <variable advanced="false" description="Specify the folder you want to access. For example, &quot;INBOX&quot; or &quot;[Gmail]/Sent Mail&quot; or any other label." group="Email Checker Parameters" hidden="false" model="PA:NOT_EMPTY_STRING" name="EMAIL_FOLDER_NAME" value="Free"/>
+    <variable advanced="false" description="Specify the number of last emails to be checked. Set -1 to check all emails." group="Email Checker Parameters" hidden="false" model="PA:INTEGER" name="CHECK_LAST_N_EMAILS" value="1"/>
+  </variables>
+  <description>
+    <![CDATA[ A workflow that accesses the specified folder in the mailbox to review each email, saving in a list of HashMap only the header details—such as the "From", "To", "Date", "Has_Attachment", and "Subject" fields—without reading the message content. ]]>
```
Comment (fviale) [inline EmailTools/resources/catalog/Imap_Email_Checker.xml:13]: A workflow that reads emails from a mail folder and extract their headers information.
Results are presented as a list of headers info containing entries such as "From", "To", "Date", "Has_Attachment", and "Subject", etc.

File: EmailTools/resources/catalog/Imap_Email_Checker.xml [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+  </description>
+  <genericInformation>
+<info name="bucketName" value="it-ci-cd"/>
+<info name="workflow.icon" value="/automation-dashboard/styles/patterns/img/wf-icons/email_checker.png"/>
+<info name="group" value="public-objects"/>
+</genericInformation>
+  <taskFlow>
+    <task fork="true" name="Imap_Email_Checker" preciousResult="true">
+      <description>
+        <![CDATA[ A workflow that accesses the specified folder in the mailbox to review each email, saving in a list of HashMap only the header details—such as the "From", "To", "Date", "Has_Attachment", and "Subject" fields—without reading the message content. ]]>
```
Comment (fviale) [inline EmailTools/resources/catalog/Imap_Email_Checker.xml:23]: see description suggestion above

File: EmailTools/resources/catalog/Imap_Email_Fetcher.xml [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+    <variable description="Email app password" group="SMTP Server Connection" model="PA:CREDENTIAL" name="PASSWORD" value="imap://${USERNAME}@${HOST}"/>
+    <variable description="Specify the folder you want to access in the mailbox. For example, &quot;INBOX&quot; or &quot;[Gmail]/Sent Mail&quot; or any other label." group="Email Fetcher Parameters" model="PA:NOT_EMPTY_STRING" name="EMAIL_FOLDER_NAME" value="EDF"/>
+    <variable description="If true, fetch the email content and save it in a html file." group="Email Fetcher Parameters" model="PA:BOOLEAN" name="FETCH_EMAIL_CONTENT" value="true"/>
+    <variable description="A comma-separated string specifying the indices of emails to fetch. Leave empty to fetch all emails." group="Email Fetcher Parameters" name="EMAIL_INDEXES" value=""/>
+    <variable description="Relative path of a folder inside the USER Data Space in which the attachment files will be saved. The provided path can be either a path to a folder terminated by / or an empty value for the root (e.g. attachments/." group="Email Fetcher Parameters" model="PA:REGEXP(^$|^.+/$)" name="DROP_FOLDER_PATH" value="/tmp/attachments/"/>
```
Comment (fviale) [inline EmailTools/resources/catalog/Imap_Email_Fetcher.xml:12]: The description mentions a relative path, but the default value is an absolute path.

File: EmailTools/resources/catalog/Imap_Email_Fetcher.xml [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+    <variable description="If true, the fetched emails will be deleted." group="Email Fetcher Parameters" model="PA:BOOLEAN" name="ENABLE_EMAIL_DELETION" value="false"/>
+  </variables>
+  <description>
+    <![CDATA[ A workflow that accesses a specified mailbox folder, retrieves the email content and attachments for the given indexes, and saves them to a designated drop folder. This workflow can also be configured to automatically delete the processed emails. ]]>
```
Comment (fviale) [inline EmailTools/resources/catalog/Imap_Email_Fetcher.xml:16]: A workflow that reads emails from a mail folder and extract their contents, including attachments.

A list of **MsgNum** indexes (see _Imap_Email_Checker_) should be provided to retrieve only specific emails.

Result is saved as a zip archive containing all emails and attachments, both as result of the task and in a dedicated folder inside the **User Space**.

File: EmailTools/resources/catalog/Imap_Email_Checker.xml (hunk 3 above)
Comment (fviale) [reply EmailTools/resources/catalog/Imap_Email_Checker.xml:23]: replace workflow with task here also

File: EmailTools/resources/catalog/Imap_Email_Fetcher.xml (hunk 4 above)
Comment (fviale) [reply EmailTools/resources/catalog/Imap_Email_Fetcher.xml:12]: @aminelouati you didn't check this comment

File: EmailTools/resources/catalog/Imap_Email_Checker.xml [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    <variable advanced="false" description="Email app password" group="SMTP Server Connection" hidden="false" model="PA:CREDENTIAL" name="PASSWORD" value="imap://${USERNAME}@${HOST}"/>
+    <variable advanced="false" description="Specify the folder you want to access. For example, &quot;INBOX&quot; or &quot;[Gmail]/Sent Mail&quot; or any other label." group="Email Checker Parameters" hidden="false" model="PA:NOT_EMPTY_STRING" name="EMAIL_FOLDER_NAME" value="Free"/>
+    <variable advanced="false" description="Specify the number of last emails to be checked. Set -1 to check all emails." group="Email Checker Parameters" hidden="false" model="PA:INTEGER" name="CHECK_LAST_N_EMAILS" value="1"/>
+  </variables>
+  <description>
+    <![CDATA[ A workflow that accesses the specified folder in the mailbox to review each email, saving in a list of HashMap only the header details—such as the **From**, **To**, **Date**, **Has\_Attachment**, **Subject**, and **MsgNum**, etc. MsgNum contains the email index, which can be used when calling the Imap_Email_Fetcher task. ]]>
```
Comment (fviale) [inline EmailTools/resources/catalog/Imap_Email_Checker.xml:13]: header details—such as
=>
header details, such as

File: EmailTools/resources/catalog/Imap_Email_Checker.xml [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+  </description>
+  <genericInformation>
+<info name="bucketName" value="it-ci-cd"/>
+<info name="workflow.icon" value="/automation-dashboard/styles/patterns/img/wf-icons/email_checker.png"/>
+<info name="group" value="public-objects"/>
+</genericInformation>
+  <taskFlow>
+    <task fork="true" name="Imap_Email_Checker" preciousResult="true">
+      <description>
+        <![CDATA[ A task that accesses the specified folder in the mailbox to review each email, saving in a list of HashMap only the header details—such as the **From**, **To**, **Date**, **Has\_Attachment**, **Subject**, and **MsgNum**, etc. MsgNum contains the email index, which can be used when calling the Imap_Email_Fetcher task. ]]>
```
Comment (fviale) [inline EmailTools/resources/catalog/Imap_Email_Checker.xml:23]: see change above

File: EmailTools/resources/catalog/Imap_Email_Fetcher.xml [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+    <variable description="Specify the folder you want to access in the mailbox. For example, &quot;INBOX&quot; or &quot;[Gmail]/Sent Mail&quot; or any other label." group="Email Fetcher Parameters" model="PA:NOT_EMPTY_STRING" name="EMAIL_FOLDER_NAME" value="Inbox"/>
+    <variable description="If true, fetch the email content and save it in a html file." group="Email Fetcher Parameters" model="PA:BOOLEAN" name="FETCH_EMAIL_CONTENT" value="true"/>
+    <variable description="A comma-separated string specifying the indices of emails to fetch. Leave empty to fetch all emails." group="Email Fetcher Parameters" name="EMAIL_INDEXES" value=""/>
+    <variable description="An absolute path or a relative path of a folder inside the USER Data Space in which the attachment files will be saved. The provided path can be either a path to a folder terminated by / or an empty value for the root (e.g. /tmp/attachments/ attachments/)." group="Email Fetcher Parameters" model="PA:REGEXP(^$|^.+/$)" name="DROP_FOLDER_PATH" value="/tmp/attachments/"/>
```
Comment (fviale) [inline EmailTools/resources/catalog/Imap_Email_Fetcher.xml:12]: ok that it can be an absolute or relative path, but I am not sure about this default value, it looks like a linux path

File: EmailTools/resources/catalog/Imap_Email_Fetcher.xml [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+    <variable description="If true, the fetched emails will be deleted." group="Email Fetcher Parameters" model="PA:BOOLEAN" name="ENABLE_EMAIL_DELETION" value="false"/>
+  </variables>
+  <description>
+    <![CDATA[ A workflow that reads emails from a mail folder and extract their contents, including attachments. A list of **MsgNum** indexes (see Imap\_Email\_Checker workflow) should be provided to retrieve only specific emails. Result is saved as a zip archive containing all emails and attachments, both as result of the task and in a dedicated folder inside the User Space. This workflow can also be configured to automatically delete the read emails.]]>
```
Comment (fviale) [inline EmailTools/resources/catalog/Imap_Email_Fetcher.xml:16]: to automatically delete the read emails
=>
to automatically delete the fetched emails

File: EmailTools/resources/catalog/Imap_Email_Fetcher.xml (hunk 8 above)
Comment (aminelouati) [reply EmailTools/resources/catalog/Imap_Email_Fetcher.xml:12]: OK

File: EmailTools/resources/catalog/Imap_Email_Fetcher.xml (hunk 9 above)
Comment (aminelouati) [reply EmailTools/resources/catalog/Imap_Email_Fetcher.xml:16]: OK

File: EmailTools/resources/catalog/Imap_Email_Fetcher.xml [hunk 10]
```diff
@@ -9,11 +9,11 @@
...
     <variable description="If true, fetch the email content and save it in a html file." group="Email Fetcher Parameters" model="PA:BOOLEAN" name="FETCH_EMAIL_CONTENT" value="true"/>
     <variable description="A comma-separated string specifying the indices of emails to fetch. Leave empty to fetch all emails." group="Email Fetcher Parameters" name="EMAIL_INDEXES" value=""/>
-    <variable description="An absolute path or a relative path of a folder inside the USER Data Space in which the attachment files will be saved. The provided path can be either a path to a folder terminated by / or an empty value for the root (e.g. /tmp/attachments/ attachments/)." group="Email Fetcher Parameters" model="PA:REGEXP(^$|^.+/$)" name="DROP_FOLDER_PATH" value="/tmp/attachments/"/>
+    <variable description="An absolute path or a relative path of a folder inside the USER Data Space in which the attachment files will be saved. The provided path can be either a path to a folder terminated by / or an empty value for the root (e.g. attachments/)." group="Email Fetcher Parameters" model="PA:REGEXP(^$|^.+/$)" name="DROP_FOLDER_PATH" value="/tmp/attachments/"/>
```
Comment (fviale) [inline EmailTools/resources/catalog/Imap_Email_Fetcher.xml:12]: you changed the description but not the default value ...

I suggest
attachments/$PA_JOB_ID/

Comment (aminelouati) [reply EmailTools/resources/catalog/Imap_Email_Fetcher.xml]: I have changed it in Try platform for testing and forget to do the same in the project

**Review by fviale — APPROVED** [review]
thanks @aminelouati
